stdlib/path.gr

```diff
@@ -0,0 +1,571 @@
+ * @returns A `PathWithType` wrapping the path with its type
+ * 
+ * @since v0.5.5
+ */
+export let wrapWithType = (path: Path<a, b>) => {
+  match (path) {
+    (Abs(_), File, _) as p => AbsoluteFile(p),
+    (Abs(_), Dir, _) as p => AbsoluteDir(p),
+    (Rel(_), File, _) as p => RelativeFile(p),
+    (Rel(_), Dir, _) as p => RelativeDir(p),
+  }
+}
+
+/**
+ * Creates a new path from a path with a relative path segment appended to it.
+ * 
+ * @param path: The base path to append the segment to
+ * @param toAppend: The relative path to append
+ * @returns A new path with the `toAppend` path appended to the `path`
+ * 
+ * @since v0.5.5
+ */
+export let rec append =
+  (
+    path: Path<a, Directory>,
+    toAppend: Path<Relative, b>,
+  ) =>
+  match (toAppend) {
+    (Rel(up2), fileType, s2) =>
+      match (path) {
+        (Rel(up1), _, []) => (Rel(up1 + up2), fileType, s2),
+        (Abs(_) as d, _, []) => (d, fileType, s2),
+        (d, pft, [_, ...rest] as s1) => {
+          if (up2 > 0) append((d, pft, rest), (Rel(up2 - 1), fileType, s2))
+          else (d, fileType, List.append(s2, s1))
+        },
+      },
+    (Abs(_), _, _) => fail "Impossible: relative path encoded as absolute path",
+  }: Path<a, b>
+
+enum DirsUp {
+  Zero,
+  Positive,
+}
+
+let dirsUp = x => if (x == 0) Zero else Positive
+
+// helper function for relativizing paths; handles the correct number of
+// directories to "go up" from one path to another
+let rec relativizeDepth = ((up1, s1), (up2, s2)) =>
+  match ((dirsUp(up1), dirsUp(up2), s1, s2)) {
+    (Zero, Zero, [hd1, ...tl1], [hd2, ...tl2]) when hd1 == hd2 =>
+      relativizeDepth((0, tl1), (0, tl2)),
+    (Zero, Zero, [], _) => Ok((up2, s2)),
+    (Zero, Zero, _, _) => Ok((List.length(s1), s2)),
+    (Positive, Positive, _, _) => relativizeDepth((up1 - 1, s1), (up2 - 1, s2)),
+    (Zero, Positive, _, _) => Ok((List.length(s1) + up2, s2)),
+    (Positive, Zero, _, _) => Err(ImpossibleRelativization),
+  }
+
+/**
+ * Attempts to construct a new relative path which will lead to the destination
+ * path from the source path. If the source and destination are both absolute
+ * paths but are on different roots, `Err(Path.DifferentRoots)` will be
+ * returned, and if the route to the destination cannot be concretely
+ * determined from the source, `Err(Path.ImpossibleRelativization)` will be
+ * returned.
+ * 
+ * @param source: The source path to find a path from
+ * @param dest: The destination path to find the path to
+ * @returns An `Ok` variant containing the relative path if successfully parsed or `Err` otherwise
+ * 
+ * @example // relativeTo(/usr, /usr/bin) => ./bin
+ * @example // relativeTo(/home/me, /home/me) => .
+ * @example // relativeTo(/usr/bin, /etc) => ../../etc
+ * @example // relativeTo(.., ../../thing) => ../thing
+ * @example // relativeTo(/usr/bin, C:/Users) => Err(DifferentRoots)
+ * @example // relativeTo(../here, ./there) => Err(ImpossibleRelativization)
```

Comment:
This isn't technically legal since the arguments should be `Path` objects and not strings so I'm thinking this might mislead potential users; Doing it the "proper" way results in somewhat long lines but maybe that's not the worst thing

Reply:
Oh, you are right. That makes this really difficult. Maybe we exclude these until we have multi-line examples? (ref #971)

Reply:
If I remove the module names it's actually not _too_ bad, the lengths become more manageable imo

e.g. `relativeTo(unwrap(absoluteDir("/usr")), unwrap(absoluteDir("/usr/bin"))) // a Path containing "./bin"`

Reply:
Going back to the idea I provided in another comment. If these all operated on the `PathWithType`, then these examples turn into `relativeTo(fromString("/usr"), fromString("/usr/bin"))` which looks pretty solid